File: src/756.py
import sys
import logging
from typing import List, Optional
from collections import defaultdict, deque, Counter
from functools import cache, lru_cache
from bisect import bisect_left, bisect_right
from itertools import accumulate
from heapq import heapify, heappush, heappop


def logging_setting():
    LOG_LEVEL = logging.INFO    # DEBUG, INFO, WARNING, ERROR, CRITICAL
    if __debug__:
        LOG_LEVEL = logging.DEBUG

    logging.basicConfig(
        stream=sys.stderr,
        level=LOG_LEVEL,
        format="%(levelname)-6s %(asctime)s %(module)s:%(funcName)s:%(lineno)-4s %(message)s",
        datefmt='%Y/%m/%d %H:%M:%S')


# https://leetcode.cn/problems/pyramid-transition-matrix/solutions/3866474/tu-jie-hui-su-vis-you-hua-pythonjavacgo-9kcb9/
class Solution:
    def __init__(self):
        self.groups = None
        self.pyramid = None
        self.vis = None

    def dfs(self, i: int, j: int) -> bool:
        retVal = False

        if i < 0:
            retVal = True
        elif self.pyramid[i] in self.vis:
            retVal = False
        elif j == i + 1:
            self.vis.add(self.pyramid[i])
            retVal = self.dfs(i - 1, 0)
        else:
            for top in self.groups[self.pyramid[i + 1] >> (j * 3) & 7][self.pyramid[i + 1] >> ((j + 1) * 3) & 7]:
                self.pyramid[i] &= ~(7 << (j * 3))
                self.pyramid[i] |= top << (j * 3)
                if self.dfs(i, j + 1):
                    retVal = True
                    break

        return retVal

    def pyramidTransition(self, bottom: str, allowed: List[str]) -> bool:
        retVal = False

        self.groups = [[[] for _ in range(7)] for _ in range(7)]
        for a, b, c in allowed:
            self.groups[ord(a) & 31][ord(b) & 31].append(ord(c) & 31)

        bottomSize = len(bottom)
        self.pyramid = [0] * bottomSize
        for i, ch in enumerate(bottom):
            self.pyramid[-1] |= (ord(ch) & 31) << (i * 3)

        self.vis = set()

        retVal = self.dfs(bottomSize - 2, 0)

        return retVal


if __name__ == "__main__":
    logging_setting()

    try:
        logging.info("sys.version: %s", sys.version)
        print()

        pSolution = Solution()
        for bottom, allowed in [["BCD", ["BCC", "CDE", "CEA", "FFF"]],
                                ["AAAA", ["AAB", "AAC", "BCD", "BBE", "DEF"]]]:
            # /* Example
            #  *  Input: bottom = "BCD", allowed = ["BCC","CDE","CEA","FFF"]
            #  *  Output: true
            #  *
            #  *  Input: bottom = "AAAA", allowed = ["AAB","AAC","BCD","BBE","DEF"]
            #  *  Output: false
            #  */
            logging.info("Input: bottom = \"%s\", allowed = %s", bottom, allowed)

            retVal = pSolution.pyramidTransition(bottom, allowed)
            logging.info("Output: %s", retVal)

            print()
    except KeyboardInterrupt as exception:
        logging.error("%s: %s", exception.__class__.__name__, exception, exc_info=True)
        pass
    except Exception as exception:
        logging.error("%s: %s", exception.__class__.__name__, exception, exc_info=True)
        pass

    sys.exit(0)

File: src/756.c
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

// https://leetcode.cn/problems/pyramid-transition-matrix/solutions/3866474/tu-jie-hui-su-vis-you-hua-pythonjavacgo-9kcb9/
#define MAXN (20)
#define MAXGROUP (7)
#define MAXTOPS (10)
bool dfs(int groups[MAXGROUP][MAXGROUP][MAXTOPS], int groupSize[MAXGROUP][MAXGROUP], int* pyramid, unsigned int* vis,
         int i, int j) {
    bool retVal = false;

    if (i < 0) {
        retVal = true;
    } else if (vis[pyramid[i]]) {
        retVal = false;
    } else if (j == i + 1) {
        vis[pyramid[i]] = 1;
        retVal = dfs(groups, groupSize, pyramid, vis, i - 1, 0);
    } else {
        int left = (pyramid[i + 1] >> (j * 3)) & 7;
        int right = (pyramid[i + 1] >> ((j + 1) * 3)) & 7;
        for (int k = 0; k < groupSize[left][right]; k++) {
            int top = groups[left][right][k];
            pyramid[i] &= ~(7 << (j * 3));
            pyramid[i] |= top << (j * 3);
            if (dfs(groups, groupSize, pyramid, vis, i, j + 1)) {
                retVal = true;
                break;
            }
        }
    }

    return retVal;
}
bool pyramidTransition(const char* bottom, char** allowed, int allowedSize) {
    bool retVal = false;

    int groups[MAXGROUP][MAXGROUP][MAXTOPS];
    int groupSize[MAXGROUP][MAXGROUP];
    memset(groups, 0, sizeof(groups));
    memset(groupSize, 0, sizeof(groupSize));

    for (int idx = 0; idx < allowedSize; idx++) {
        char* s = allowed[idx];
        int a = s[0] & 31;
        int b = s[1] & 31;
        int c = s[2] & 31;
        groups[a][b][groupSize[a][b]++] = c;
    }

    int bottomSize = strlen(bottom);
    int pyramid[MAXN] = {0};
    for (int i = 0; i < bottomSize; i++) {
        pyramid[bottomSize - 1] |= (bottom[i] & 31) << (i * 3);
    }

    unsigned int* vis = (unsigned int*)calloc(1 << ((bottomSize - 1) * 3), sizeof(unsigned int));

    retVal = dfs(groups, groupSize, pyramid, vis, bottomSize - 2, 0);

    //
    free(vis);

    return retVal;
}

int main(int argc, char** argv) {
#define MAX_SIZE (216)
    struct testCaseType {
        char* bottom;
        char* allowed[MAX_SIZE];
        int allowedSize;
    } testCase[] = {{"BCD", {"BCC", "CDE", "CEA", "FFF"}, 4}, {"AAAA", {"AAB", "AAC", "BCD", "BBE", "DEF"}, 5}};
    int numberOfTestCase = sizeof(testCase) / sizeof(testCase[0]);
    /* Example
     *  Input: bottom = "BCD", allowed = ["BCC","CDE","CEA","FFF"]
     *  Output: true
     *
     *  Input: bottom = "AAAA", allowed = ["AAB","AAC","BCD","BBE","DEF"]
     *  Output: false
     */

    bool answer;
    int i, j;
    for (i = 0; i < numberOfTestCase; ++i) {
        printf("Input: bottom = %s, allowed = [", testCase[i].bottom);
        for (j = 0; j < testCase[i].allowedSize; ++j) {
            printf("%s%s", (j == 0) ? "" : ",", testCase[i].allowed[j]);
        }
        printf("]\n");

        answer = pyramidTransition(testCase[i].bottom, testCase[i].allowed, testCase[i].allowedSize);
        printf("Output: %s\n", (answer == true ? "true" : "false"));

        printf("\n");
    }

    return EXIT_SUCCESS;
}
